ksqldb-engine/src/main/java/io/confluent/ksql/execution/ExecutionPlanBuilder.java

```diff
@@ -87,4 +97,142 @@ public ExecutionPlan buildPhysicalPlan(
         resultStream.getSourceStep()
     );
   }
+
+
+  private void optimizeForSelfJoin(final PlanNode currentNode) {
+
+    if (currentNode instanceof JoinNode) {
+      final JoinNode joinNode = (JoinNode) currentNode;
+      if (! matchSubGraphPattern1(joinNode)) {
+        if (!matchSubGraphPattern2(joinNode)) {
+          matchSubGraphPattern3(joinNode);
+        }
+      }
+    }
+
+    for (final PlanNode child: currentNode.getSources()) {
+      optimizeForSelfJoin(child);
+    }
+  }
+
+  /**
+   * Join -> PreJoinProject -> PreJoinRepartition -> DataSource
+   */
+  // CHECKSTYLE_RULES.OFF: CyclomaticComplexity
+  private boolean matchSubGraphPattern1(final JoinNode joinNode) {
+    // CHECKSTYLE_RULES.ON: CyclomaticComplexity
+    if (joinNode.getLeft() instanceof PreJoinProjectNode
+        && joinNode.getRight() instanceof PreJoinProjectNode) {
+      final PreJoinProjectNode preJoinProjectNodeLeft =
+          (PreJoinProjectNode) joinNode.getLeft();
+      final PreJoinProjectNode preJoinProjectNodeRight =
+          (PreJoinProjectNode) joinNode.getRight();
+      if (preJoinProjectNodeLeft.getSource() instanceof PreJoinRepartitionNode
+          && preJoinProjectNodeRight.getSource() instanceof PreJoinRepartitionNode) {
+
+        final PreJoinRepartitionNode preJoinRepartitionNodeLeft =
+            (PreJoinRepartitionNode) preJoinProjectNodeLeft.getSource();
+        final PreJoinRepartitionNode preJoinRepartitionNodeRight =
+            (PreJoinRepartitionNode) preJoinProjectNodeRight.getSource();
+
+        // Check if Repartition is on the same key
+        if (!preJoinRepartitionNodeLeft.getPartitionBy()
+            .equals(preJoinRepartitionNodeRight.getPartitionBy())) {
+          return false;
+        }
+
+        // Check if sources are the same stream
+        if (!(preJoinRepartitionNodeLeft.getSource() instanceof DataSourceNode)
+            || !(preJoinRepartitionNodeRight.getSource() instanceof DataSourceNode)) {
+          return false;
+        }
+        final DataSourceNode leftSource = (DataSourceNode) preJoinRepartitionNodeLeft.getSource();
+        final DataSourceNode rightSource = (DataSourceNode) preJoinRepartitionNodeRight.getSource();
+
+        if (leftSource.getDataSource().getDataSourceType().equals(DataSourceType.KSTREAM)
+            && rightSource.getDataSource().getDataSourceType().equals(DataSourceType.KSTREAM)) {
+
+          if (leftSource.getDataSource().getName().equals(rightSource.getDataSource().getName())) {
+
+            // Annotate join as self-join
+            joinNode.setSelfJoin(true);
+            preJoinProjectNodeLeft.setSelfJoin(true);
+            preJoinProjectNodeRight.setSelfJoin(true);
```

Comment:
It's not possible to do while building the logical plan as the optimizations need to see the whole plan to be able to determine its feasibility. The alternative would be to add an optimization phase (after the logical plan is built and before the physical plan is built) that would traverse the logical plan and apply optimizations. I didn't go with this approach because:
1. It's not really changing the plan besides setting a flag. 
2. We are planning on changing the logical--> physical plan so it would be throw-away work
3. This seemed the fastest, easiest and less obtrusive way to achieve this.